control-plane/typeSpecConverter.cpp

```diff
@@ -111,6 +115,49 @@ bool TypeSpecConverter::preorder(const IR::Type_Name* type) {
     return false;
 }
 
+bool TypeSpecConverter::preorder(const IR::Type_Newtype* type) {
+    if (p4RtTypeInfo) {
+        bool orig = true;
+        cstring uri;
+        uint32_t sdnB;
+        auto ann = type->getAnnotation(IR::Annotation::p4runtimeTranslationAnnotation);
+        if (ann != nullptr) {
+            orig = false;
+            int i = 0;
+            for (auto a : ann->body) {
```

Comment:
I don't think this code makes much sense any more now that you are parsing the annotation with `PARSE_PAIR`. The whole point of the `parseAnnotations` pass is that you *parse* the annotations. You should be able to use `IR::Annotation::expr` now as per the documentation (https://github.com/p4lang/p4c/blob/master/ir/base.def#L209). They are a lot of examples all over the place as well.

Reply:
If you see the P4 test program in `testdata/p4_16_samples/p416-type-use3.p4`, many `type`s do not have an annotation.  Thus, the compiler crashes unless the code has `if (ann != nullptr)`.  The compiler has to cater to a user who forgot to add an annotation.

The `@p4runtime_translation` annotation is parsed by the compiler as an AnnotationToken vector, not an Expression vector.  I will see what I can do to make the code brief.

Reply:
Regarding your comment "The compiler has to cater to a user who forgot to add an annotation".

I am not saying that I know there is anything wrong with your code changes.  I am commenting on the statement of yours that I quoted, because it sounds like a misunderstanding to me.

It is 100% legal that a P4_16 program author can write a program with a `type` definition, without a `@p4runtime_translation` annotation on it.  They may have forgotten to add such an annotation, but it is also possible that they intended to do this -- e.g. perhaps they wanted a new `type` name because they want to disallow automatic assignments between values of the new type and its base type without an explicit cast, because they think it will help the compiler catch them in making certain kinds of errors in their P4 program.

Reply:
Whether a 'type` has been annotated or not, p4c will disallow automatic assignment without a Cast.  Bottom-line is that p4c should not crash with any P4 code.